lib/deepcrawl.js

```diff
@@ -0,0 +1,137 @@
+'use strict';
+const BPromise = require('bluebird'),
+  schema = require('../schemas/v2.0.0'),
+  needle = BPromise.promisifyAll(require('needle')),
+  humps = require('humps'),
+  _ = require('lodash');
+
+class DeepCrawl {
+  constructor(cfg) {
+    cfg = cfg || {};
+
+    this.apiId = cfg.apiId;
+    this.apiKey = cfg.apiKey;
+    this.getSessionToken = cfg.getSessionToken;
+    this.accountId = cfg.accountId;
+
+    this.schema = cfg.schema || schema;
+    this.version = this.schema.version;
+    this.baseUrl = this.schema.baseUrl;
+  }
+
+  createEndpoints (schema) {
+    const API = {};
+    for (const resource in schema.resources) {
+      API[resource] = {};
+    }
+    return API;
+  }
+
+  generateRequest (url, options, method) {
+    // first, get the sessionToken and generate the request options and headers
+    return BPromise.resolve(options.sessionToken || this.getSessionToken())
+      .then((sessionToken) => {
+        const requestOpts = {
+          json: true,
+          headers: {
+            'X-Auth-Token': sessionToken
+          }
+        };
+
+        if (method === 'get') {
+          return needle.getAsync(url, requestOpts)
+            .then(this.handleResponse);
+        }
+
+        // convert all camelCased options to _ for deepcrawl
+        const data = humps.decamelizeKeys(options);
+        return needle[`${method}Async`](url, data, requestOpts)
+          .then(this.handleResponse);
+      });
+  }
+
+  generateMethod (requiredFields, url, method) {
+    return (options) => {
+      options = options || {};
+      if (!options.sessionToken && !this.getSessionToken) {
+        throw new Error(`"sessionToken" or "getSessionToken" method required`);
+      }
+
+      requiredFields.forEach((field) => {
+        if (!options[field] && !this[field]) {
+          throw new Error(`${field} is required`);
+        }
+        // if the url contains {field}, replace {field} with the passed
+        // in value
+        if (url.indexOf(`{${field}}`) > -1) {
+          url = url.replace(`{${field}}`, (options[field] || this[field]));
+        }
+      });
+
+      return this.generateRequest(url, options, method);     
+    };
+  }
+
+  addMethodsToEndpoints (API, schema) {
+    for (const resource in API) {
+      const resourceUrl = this.baseUrl + schema.resources[resource].route,
+        requiredFields = _.clone(schema.resources[resource].requiredFields || []),
+        id = schema.resources[resource].id;
+
+      for (const action in schema.resources[resource].actions) {
+        let actionRequiredFields = requiredFields,
+          actionUrl = resourceUrl;
+        // check to see whether the resource action has extra data associated with it
+        const isString = typeof schema.resources[resource].actions[action] === 'string';
```

Comment:
In fact, instead of this `isString` var, a more elegant approach (to avoid the conditional checking below) is to convert the string into a defaulted object, e.g.:
```
  if (typeof action === 'string') {
    action = {
      type: action,
      requiredFields = []
    };
  };
```
It's slightly more code here, but then less (and much simpler) code later.